[SPARK-9782] [YARN] Support YARN application tags via SparkConf

dennishuo opened the pull request:
Add a new test case in yarn/ClientSuite which checks how the various SparkConf
and ClientArguments propagate into the ApplicationSubmissionContext.

dennishuo committed:
[SPARK-9782] [YARN] Support YARN application tags via SparkConf

Add a new test case in yarn/ClientSuite which checks how the various SparkConf
and ClientArguments propagate into the ApplicationSubmissionContext.

AmplabJenkins commented:
Can one of the admins verify this patch?

tgravescs commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

sryza commented on docs/running-on-yarn.md:
```diff
@@ -320,6 +320,14 @@ If you need a reference to the proper location to put log files in the YARN so t
+  <td>(none)</td>
+  <td>
+  Comma-separated list of strings to pass through as YARN application tags appearing
+  in YARN ApplicationReports, which can be used for filtering when querying YARN.
```
"when querying YARN apps"

dennishuo replied:
Done.

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -163,6 +163,23 @@ private[spark] class Client(
+      .filter(!_.isEmpty())
+      .foreach { tagCollection =>
+        try {
+          val method = appContext.getClass().getMethod(
```
Add a comment here mentioning why this needs to be called via reflection

dennishuo replied:
Done.

sryza commented on docs/running-on-yarn.md:
```diff
@@ -320,6 +320,14 @@ If you need a reference to the proper location to put log files in the YARN so t
   </td>
 </tr>
 <tr>
+  <td><code>spark.yarn.application.tags</code></td>
```
I would change this to simply "spark.yarn.tags"

dennishuo replied:
Done.

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -163,6 +163,23 @@ private[spark] class Client(
+          method.invoke(appContext, new java.util.HashSet[String](tagCollection))
+          logInfo("Applied setApplicationTags based on %s='%s'"
+            .format(CONF_SPARK_YARN_APPLICATION_TAGS, tagCollection))
+
```
Extra space unnecessary

dennishuo replied:
Done.

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -163,6 +163,23 @@ private[spark] class Client(
+          val method = appContext.getClass().getMethod(
+            "setApplicationTags", classOf[java.util.Set[String]])
+          method.invoke(appContext, new java.util.HashSet[String](tagCollection))
+          logInfo("Applied setApplicationTags based on %s='%s'"
```
I'd omit this log message.  I don't think it's more important than other attributes like queue and app name, which we don't log.

dennishuo replied:
Done.

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -163,6 +163,23 @@ private[spark] class Client(
+
+        } catch {
+          case e: NoSuchMethodException =>
+            logWarning("Ignoring conf %s='%s'; setApplicationTags missing in this version of YARN."
```
I'd say "Ignoring spark.yarn.tags because this version of YARN does not support it"

dennishuo replied:
Done.

sryza commented on yarn/src/test/scala/org/apache/spark/deploy/yarn/ClientSuite.scala:
```diff
@@ -170,6 +173,39 @@ class ClientSuite extends SparkFunSuite with Matchers with BeforeAndAfterAll {
     cp should contain ("/remotePath/my1.jar")
   }
 
+  test("configuration and args propagate through createApplicationSubmissionContext") {
```
Have you verified that this test passes both against versions of YARN that support app tags and against versions that do not?

dennishuo replied:
Indeed I did before sending out the first PR:

$ sbt/sbt -Pyarn -Phadoop-2.3 -Phive -Phive-thriftserver yarn/test test-only 
...
[info] Compiling 1 Scala source to /home/dhuo/github/dhuo_spark2/spark/yarn/target/scala-2.10/test-classes...
[info] YarnAllocatorSuite:
[info] - single container allocated (667 milliseconds)
[info] - some containers allocated (221 milliseconds)
[info] - receive more containers than requested (200 milliseconds)
[info] - decrease total requested executors (17 milliseconds)
[info] - decrease total requested executors to less than currently running (13 milliseconds)
[info] - kill executors (25 milliseconds)
[info] - lost executor removed from backend (26 milliseconds)
[info] - memory exceeded diagnostic regexes (5 milliseconds)
[info] YarnClusterSuite:
[info] - run Spark in yarn-client mode (14 seconds, 642 milliseconds)
[info] - run Spark in yarn-cluster mode (13 seconds, 292 milliseconds)
[info] - run Spark in yarn-cluster mode unsuccessfully (10 seconds, 150 milliseconds)
[info] - run Python application in yarn-client mode (16 seconds, 129 milliseconds)
[info] - run Python application in yarn-cluster mode (17 seconds, 323 milliseconds)
[info] - user class path first in client mode (13 seconds, 628 milliseconds)
[info] - user class path first in cluster mode (13 seconds, 120 milliseconds)
[info] ContainerPlacementStrategySuite:
[info] - allocate locality preferred containers with enough resource and no matched existed containers (838 milliseconds)
[info] - allocate locality preferred containers with enough resource and partially matched containers (21 milliseconds)
[info] - allocate locality preferred containers with limited resource and partially matched containers (16 milliseconds)
[info] - allocate locality preferred containers with fully matched containers (14 milliseconds)
[info] - allocate containers with no locality preference (16 milliseconds)
[info] YarnSparkHadoopUtilSuite:
[info] - shell script escaping (17 milliseconds)
[info] - Yarn configuration override (53 milliseconds)
[info] - test getApplicationAclsForYarn acls on (10 milliseconds)
[info] - test getApplicationAclsForYarn acls on and specify users (12 milliseconds)
[info] - test expandEnvironment result (4 milliseconds)
[info] - test getClassPathSeparator result (6 milliseconds)
[info] - check access nns empty (20 milliseconds)
[info] - check access nns unset (18 milliseconds)
[info] - check access nns (16 milliseconds)
[info] - check access nns space (17 milliseconds)
[info] - check access two nns (16 milliseconds)
[info] - check token renewer (206 milliseconds)
[info] - check token renewer default (32 milliseconds)
[info] ClientSuite:
[info] - default Yarn application classpath (29 milliseconds)
[info] - default MR application classpath (3 milliseconds)
[info] - resultant classpath for an application that defines a classpath for YARN (16 milliseconds)
[info] - resultant classpath for an application that defines a classpath for MR (11 milliseconds)
[info] - resultant classpath for an application that defines both classpaths, YARN and MR (9 milliseconds)
[info] - Local jar URIs (40 milliseconds)
[info] - Jar path propagation through SparkConf (115 milliseconds)
[info] - Cluster path translation (14 milliseconds)
[info] - configuration and args propagate through createApplicationSubmissionContext (8 milliseconds)
[info] ClientDistributedCacheManagerSuite:
[info] - test getFileStatus empty (24 milliseconds)
[info] - test getFileStatus cached (1 millisecond)
[info] - test addResource (10 milliseconds)
[info] - test addResource link null (3 milliseconds)
[info] - test addResource appmaster only (1 millisecond)
[info] - test addResource archive (7 milliseconds)
[info] ScalaTest
[info] Run completed in 2 minutes.
[info] Total number of tests run: 48
[info] Suites: completed 6, aborted 0
[info] Tests: succeeded 48, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Passed: Total 48, Failed 0, Errors 0, Passed 48
[success] Total time: 136 s, completed Aug 7, 2015 6:47:43 PM

$ sbt/sbt -Pyarn -Phadoop-2.4 -Phive -Phive-thriftserver yarn/test
...
[info] YarnAllocatorSuite:
[info] - single container allocated (1 second, 508 milliseconds)
[info] - some containers allocated (163 milliseconds)
[info] - receive more containers than requested (160 milliseconds)
[info] - decrease total requested executors (26 milliseconds)
[info] - decrease total requested executors to less than currently running (13 milliseconds)
[info] - kill executors (96 milliseconds)
[info] - lost executor removed from backend (48 milliseconds)
[info] - memory exceeded diagnostic regexes (13 milliseconds)
[info] YarnClusterSuite:
[info] - run Spark in yarn-client mode (22 seconds, 554 milliseconds)
[info] - run Spark in yarn-cluster mode (15 seconds, 772 milliseconds)
[info] - run Spark in yarn-cluster mode unsuccessfully (10 seconds, 285 milliseconds)
[info] - run Python application in yarn-client mode (18 seconds, 747 milliseconds)
[info] - run Python application in yarn-cluster mode (16 seconds, 356 milliseconds)
[info] - user class path first in client mode (13 seconds, 841 milliseconds)
[info] - user class path first in cluster mode (14 seconds, 344 milliseconds)
[info] ContainerPlacementStrategySuite:
[info] - allocate locality preferred containers with enough resource and no matched existed containers (719 milliseconds)
[info] - allocate locality preferred containers with enough resource and partially matched containers (14 milliseconds)
[info] - allocate locality preferred containers with limited resource and partially matched containers (15 milliseconds)
[info] - allocate locality preferred containers with fully matched containers (14 milliseconds)
[info] - allocate containers with no locality preference (16 milliseconds)
[info] YarnSparkHadoopUtilSuite:
[info] - shell script escaping (18 milliseconds)
[info] - Yarn configuration override (58 milliseconds)
[info] - test getApplicationAclsForYarn acls on (11 milliseconds)
[info] - test getApplicationAclsForYarn acls on and specify users (17 milliseconds)
[info] - test expandEnvironment result (9 milliseconds)
[info] - test getClassPathSeparator result (5 milliseconds)
[info] - check access nns empty (22 milliseconds)
[info] - check access nns unset (21 milliseconds)
[info] - check access nns (22 milliseconds)
[info] - check access nns space (17 milliseconds)
[info] - check access two nns (18 milliseconds)
[info] - check token renewer (176 milliseconds)
[info] - check token renewer default (82 milliseconds)
[info] ClientSuite:
[info] - default Yarn application classpath (293 milliseconds)
[info] - default MR application classpath (5 milliseconds)
[info] - resultant classpath for an application that defines a classpath for YARN (50 milliseconds)
[info] - resultant classpath for an application that defines a classpath for MR (19 milliseconds)
[info] - resultant classpath for an application that defines both classpaths, YARN and MR (16 milliseconds)
[info] - Local jar URIs (87 milliseconds)
[info] - Jar path propagation through SparkConf (172 milliseconds)
[info] - Cluster path translation (12 milliseconds)
[info] - configuration and args propagate through createApplicationSubmissionContext (62 milliseconds)
[info] ClientDistributedCacheManagerSuite:
[info] - test getFileStatus empty (30 milliseconds)
[info] - test getFileStatus cached (0 milliseconds)
[info] - test addResource (12 milliseconds)
[info] - test addResource link null (3 milliseconds)
[info] - test addResource appmaster only (1 millisecond)
[info] - test addResource archive (14 milliseconds)
[info] ScalaTest
[info] Run completed in 2 minutes, 18 seconds.
[info] Total number of tests run: 48
[info] Suites: completed 6, aborted 0
[info] Tests: succeeded 48, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Passed: Total 48, Failed 0, Errors 0, Passed 48
[success] Total time: 156 s, completed Aug 10, 2015 10:58:20 AM

Kicking off a new round of build/tests locally now after the changes.

dennishuo replied:
Tests were green with both a 2.3 and 2.4 build. Thanks for the help reviewing!

dennishuo replied:
Console output from my test runs here: https://gist.github.com/dennishuo/79212ba4f8a3bfb71227

dennishuo committed:
Minor fixes for review comments.

Rename from spark.yarn.application.tags to simply spark.yarn.tags.
Add a comment explaining the use of reflection.
Misc rewording formatting fixes.

sryza commented:
jenkins, retest this please

sryza commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -163,6 +163,23 @@ private[spark] class Client(
+          method.invoke(appContext, new java.util.HashSet[String](tagCollection))
+        } catch {
+          case e: NoSuchMethodException =>
+            logWarning("Ignoring %s='%s' because this version of YARN does not support it"
```
Nit: use `s"Ignoring $CONF_SPARK_YARN_APPLICATION_TAGS because"`...

dennishuo replied:
Done.

dennishuo committed:
Fix nit; use string interpolation for warning message instead of .format

sryza commented:
@tgravescs @vanzin @andrewor14 can I get one of you to sign off on my proposed property name `spark.yarn.tags`?

Otherwise, this LGTM

vanzin commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -163,6 +163,23 @@ private[spark] class Client(
     appContext.setQueue(args.amQueue)
     appContext.setAMContainerSpec(containerContext)
     appContext.setApplicationType("SPARK")
+    sparkConf.getOption(CONF_SPARK_YARN_APPLICATION_TAGS)
```
Not a big deal, but I'd flip this around and do something like:

    Try(appContext.getClass().getMethod("setApplicationTags", classOf[java.util.Set[String]])).foreach { setTagsMethod =>
      ...
    }

That way you print the log message only once, lookup the method only once, and skip all the processing of the option if the method is not available.

vanzin replied:
BTW if you do this you're probably better off doing:

    Try(...) match {
        case Success =>
        case Failure =>
    }

dennishuo replied:
Thanks for the suggestion! I'm a bit rusty in Scala, so I could be misapplying some idioms here, but my intent on this line was to use the foreach, etc., only to access the "None or single element" of the Option, as indicated on http://www.scala-lang.org/api/current/index.html#scala.Option , so the "foreach" here is supposed to only iterate over the "single" element which happens to itself be a collection, as opposed to iterating over elements of the inner collection. So there shouldn't be any way to cause multiple log statements or multiple reflection-based lookups of the method.

I also wanted to err on the side of minimizing behavioral changes for existing setups, so that if the Option is None, then this chaining as a monad avoids ever needing to lookup a method, or even invoking any of the option-processing methods like StringUtils.getTrimmedStringCollection.

I could add a note to make it more clear that the map/filter/foreach is on the Option, as opposed to the Collection if that'd help.

Anyhow, I'll be happy to apply the reversal to start with the method as you suggest if you prefer, just wanted to hear your thoughts on this Option usage first.

vanzin replied:
Ah, I see. `setApplicationTags` takes a `Set[String]`, not a single string, so you're not calling it once per tag in the collection. My bad. Current code is fine.

vanzin commented:
Config name LGTM. Code too, just left a minor suggestion.

tgravescs commented:
I think the name is ok, but it might be better if we have put application in it: spark.yarn.application.tags.  thoughts?
My thinking is that way if in the future some other config with tags in the name of it comes along it won't be confusing.  It matches the setApplicationTags name too.

If both of you prefer the current name I'm ok with it though.

sryza commented:
@tgravescs my thinking for just spark.yarn.tags was that it's redundant to include application, because configs are by definition per-application.  We'd also be consistent with spark.yarn.queue and spark.yarn.jars.  If we do include "application", I think it should be spark.yarn.application-tags, because then we're not adding a new "application" namespace.

tgravescs commented:
thats fine.  we can keep it tags and if another config with tags in it comes along be more specific on that.

sryza commented:
Cool, in that case @dennishuo mind making the change that @vanzin suggested and then I'll merge this?

dennishuo commented:
Thanks everyone for the reviews! @sryza it sounds like after further clarification with @vanzin we're okay with the code as-is. Is there anything else I should look at, or are you able to merge this in the current state?

shijinkui commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -163,6 +163,23 @@ private[spark] class Client(
+            logWarning(s"Ignoring $CONF_SPARK_YARN_APPLICATION_TAGS because this version of " +
+              "YARN does not support it")
+        }
+      }
```
using reflect make the code reading more difficult. generally not be recommanded.
